Refactor Autotuner

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
* Refactor C++ Autotuner implementation to work on multi-dimensional parameters using std::array
* Refactor C++ base class structure to provide a common Python interface in `AutotunedObject`
* Refactor Python base class structure to provide a common interface to
    * Get kernel parameters
    * Set kernel parameters
    * Test whether kernel parameter tuning is complete
    * Start tuning kernel parameters
* Add convenience methods to do the above for all children in `Operations`.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
* Allow users to save and restore optimal kernel parameters.
* Allow benchmarks to query when tuning is complete so they can proceed with timing measurements.
* Allow users to start a new tuning sequence when needed (e.g. after compressing the system).

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #562

## How has this been tested?

Tested in interactive sessions:

```
>>> benchmark.sim.operations.integrator.tune_kernel_parameters()
notice(4): Autotuner hpmc_moves starting scan.
notice(4): Autotuner hpmc_update_pdata starting scan.
notice(4): Autotuner hpmc_excell_block_size starting scan.
notice(4): Autotuner hpmc_num_depletants starting scan.
notice(4): Autotuner hpmc_num_depletants_ntrial starting scan.
notice(4): Autotuner hpmc_narrow starting scan.
notice(4): Autotuner hpmc_convergence starting scan.
notice(4): Autotuner hpmc_depletants_accept starting scan.
notice(4): Autotuner hpmc_depletants starting scan.
notice(4): Autotuner hpmc_depletants_phase1 starting scan.
notice(4): Autotuner hpmc_depletants_phase2 starting scan.
notice(4): Autotuner cell_list starting scan.
notice(4): Autotuner cell_list_combine starting scan.
>>> benchmark.sim.operations.integrator.is_tuning_complete
False
>>> benchmark.sim.operations.integrator.kernel_parameters
{'hpmc_moves': (32,), 'hpmc_update_pdata': (32,), 'hpmc_excell_block_size': (32,), 'hpmc_num_depletants': (32,), 'hpmc_num_depletants_ntrial': (32,), 'hpmc_narrow': (32, 1, 4), 'hpmc_convergence': (32,), 'hpmc_depletants_accept': (32,), 'hpmc_depletants': (32, 4, 4), 'hpmc_depletants_phase1': (32, 4, 4), 'hpmc_depletants_phase2': (32, 4, 4)}
>>> benchmark.sim.run(10)
notice(5): IntegratorHPMCMono: updating nominal width to 1
notice(4): Autotuner hpmc_convergence found optimal parameter (992,) with a performance spread of 18%.
>>> benchmark.sim.operations.integrator.kernel_parameters
{'hpmc_moves': (288,), 'hpmc_update_pdata': (288,), 'hpmc_excell_block_size': (352,), 'hpmc_num_depletants': (32,), 'hpmc_num_depletants_ntrial': (32,), 'hpmc_narrow': (320, 1, 2), 'hpmc_convergence': (992,), 'hpmc_depletants_accept': (32,), 'hpmc_depletants': (32, 4, 4), 'hpmc_depletants_phase1': (32, 4, 4), 'hpmc_depletants_phase2': (32, 4, 4)}
>>> benchmark.sim.operations.integrator.is_tuning_complete
False
>>> benchmark.sim.run(100)
notice(5): IntegratorHPMCMono: updating nominal width to 1
notice(4): Autotuner hpmc_narrow found optimal parameter (32,1,1,) with a performance spread of 2236%.
notice(4): Autotuner hpmc_moves found optimal parameter (32,) with a performance spread of 66%.
notice(4): Autotuner hpmc_update_pdata found optimal parameter (256,) with a performance spread of 54%.
>>> benchmark.sim.operations.integrator.kernel_parameters
{'hpmc_moves': (32,), 'hpmc_update_pdata': (256,), 'hpmc_excell_block_size': (480,), 'hpmc_num_depletants': (32,), 'hpmc_num_depletants_ntrial': (32,), 'hpmc_narrow': (32, 1, 1), 'hpmc_convergence': (992,), 'hpmc_depletants_accept': (32,), 'hpmc_depletants': (32, 4, 4), 'hpmc_depletants_phase1': (32, 4, 4), 'hpmc_depletants_phase2': (32, 4, 4)}
>>> benchmark.sim.operations.integrator.is_tuning_complete
False
>>> benchmark.sim.run(1000)
notice(5): IntegratorHPMCMono: updating nominal width to 1
notice(4): Autotuner cell_list found optimal parameter (64,) with a performance spread of 80%.
notice(4): Autotuner hpmc_excell_block_size found optimal parameter (32,) with a performance spread of 142%.
>>> benchmark.sim.operations.integrator.kernel_parameters
{'hpmc_moves': (32,), 'hpmc_update_pdata': (256,), 'hpmc_excell_block_size': (32,), 'hpmc_num_depletants': (32,), 'hpmc_num_depletants_ntrial': (32,), 'hpmc_narrow': (32, 1, 1), 'hpmc_convergence': (992,), 'hpmc_depletants_accept': (32,), 'hpmc_depletants': (32, 4, 4), 'hpmc_depletants_phase1': (32, 4, 4), 'hpmc_depletants_phase2': (32, 4, 4)}
>>> benchmark.sim.operations.integrator.is_tuning_complete
False
```

The last line should be `True`, I will investigate later. It does become `True` for the initial tuning, but stays `False` when calling `tune_kernel_parameters`.

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Added:

* `AutotunedObject` class that provides an interface to read and write tuned kernel parameters, query whether tuning is complete, and start tuning again at the object level.
* `is_tuning_complete` method to `Operations`. Check whether kernel parameter tuning is complete for all operations.
* `tune_kernel_parameters` methods to `Operations` and many other classes. Start tuning kernel parameters in all operations.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

## TODO: 
- [x] add unit tests
- [x] Write change log entry
- [x] Fix `False` `is_tuning_complete` after calling `tune_kernel_parameters`
- [x] Port MD and other components.
- [x] Add overview documentation.

joaander committed:
Convert setAutotunerParams to startAutotuner.

joaander committed:
Refactor scanning state machine.

joaander committed:
Initial testing of new tuner.

joaander committed:
Include tuple parameter in notice output.

joaander committed:
Add base Action class to manage the common autotuning API.

joaander committed:
Enable CellListGPU and LoadBalanceGPU with the new centeralized Autotuner API.

joaander committed:
Start adding Python API.

joaander committed:
Merge branch 'trunk-minor' into refactor-autotuner

joaander committed:
Add some TODO messages and remove unneeded methods.

joaander committed:
Separate out Autotuned base class from Action/Operation

joaander committed:
Add valid parameter filter.

joaander committed:
Started updating HPMC kernel autotuners.

joaander committed:
Porting more HPMC classes to the new Autotuner.

joaander committed:
Initial working version.

joaander committed:
Test get parameters.

joaander committed:
Fix bugs with is_autotuning_complete.

joaander committed:
Implement parameter get/set.

joaander committed:
Use state machine to implement optional parameters.

joaander committed:
Run clang-format and yapf.

joaander committed:
Draft documentation.

joaander committed:
Fix typo.

joaander committed:
Add TODO note.

b-butler commented on hoomd/Action.cc:
```diff
@@ -0,0 +1,22 @@
+    {
+    pybind11::class_<Action, std::shared_ptr<Action>>(m, "Action")
+        .def(pybind11::init<std::shared_ptr<SystemDefinition>>())
+        .def("getAutotunerParameters", &Action::getAutotunerParameters)
+        .def("setAutotunerParameters", &Action::setAutotunerParameters)
+        .def("startAutotuning", &Action::startAutotuning)
+        .def("isAutotuningComplete", &Action::isAutotuningComplete);
```
Why do we expose this at the `Action` and not `Autotuned` level?

joaander replied:
Because I implemented it a the `Action` level first, then later factored out the `Autotuned` class. This will need to be fixed for `Autotuned` classes (like MD integration methods) to have a functioning Python API.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+        unsigned int warp_size = exec_conf->dev_prop.warpSize;
+        if (force_size > 0)
+            {
+            warp_size = (unsigned int)force_size;
```
```suggestion
            warp_size = static_cast<unsigned int>(force_size);
```

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
-        m_exec_conf->msg->notice(6)
-            << "Set Autotuner " << m_name << " period = " << period << std::endl;
-        m_period = period;
+        if (m_state != SCANNING)
```
Since `begin` changes `INACTIVE` to `SCANNING` shouldn't that also be considered not complete?

joaander replied:
This is intentional. I want tuners that are created, but`begin` is never called to be "complete". Otherwise, they would prevent the `Autotuned` object as a whole from ever registering as complete.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+        m_n_samples += 1;
+
+    // Initialize memory.
+    m_samples.resize(m_parameters.size());
```
Should you resize `m_parameters` to possibly free up memory?

joaander replied:
We could call `shrink_to_fit`: https://en.cppreference.com/w/cpp/container/vector/shrink_to_fit

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+                float sum = 0.0f;
+                for (std::vector<float>::iterator it = v.begin(); it != v.end(); ++it)
+                    sum += *it;
+                m_sample_median[i] = sum / float(v.size());
```
Should we rename this variable considering it isn't always the median?

joaander replied:
Yes.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+
+    size_t min_idx = 0;
+
+    if (is_root)
```
```suggestion
    // Report performance characteristics of Autotuning
    if (is_root)
```

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+    elements. It then chooses the fastest time (with the lowest index breaking a tie) and returns
+    the index of the parameter that resulted in that time.
+*/
+template<size_t n_dimensions> size_t Autotuner<n_dimensions>::computeOptimalParameterIndex()
```
If possible, I think this would be cleaner as two methods: one that computes the performance metric and another that then gets the best index and reports the  #performance.

joaander replied:
Due to a lack of time, I will not implement this suggestion.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+
+#ifndef __HIPCC__
+    /// Get the autotuner parameters as a Python tuple.
+    virtual pybind11::tuple getParameterPython()
```
Wouldn't a method like this work better as a pure virtual method? The class is an interface, but we provide default (null) implementations for all the methods.

Or is this a requirement since we use this type to store autotuners in a `std::vector`?

joaander replied:
I don't know if it is a requirement of `std::vector` or not.

joaander replied:
This class really isn't an interface as it does implement some common code. I will rename it.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+                }
+            else
+                {
+                initializeParameters(dimension_ranges,
```
```suggestion
                // populate parameter for higher dimensions
                initializeParameters(dimension_ranges,
```

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,620 @@
+                m_current_param = m_parameters[m_current_element];
+                }
+            }
+        else
+            {
+            m_current_param = m_parameters[m_current_element];
+            }
```
Is this necessary?

joaander replied:
No. It is an error, in fact. It will overwrite a `m_current_param` set by the user. It is probably left over from the old `Autotuner` code which had a more complex state machine.

joaander replied:
On looking at this more carefully, the code is correct as implemented. It does not override the parameter set by the user because this is in an `if (m_state == SCANNING)` block.

joaander replied:
The reason this is necessary in the first place is because of the order in when the state machine moves over elements and samples.

b-butler commented on hoomd/CellListGPU.cc:
```diff
@@ -24,22 +24,15 @@ CellListGPU::CellListGPU(std::shared_ptr<SystemDefinition> sysdef)
-                                        100000,
-                                        "cell_list_combine",
-                                        this->m_exec_conf));
+    m_tuner.reset(new Autotuner<1>({AutotunerInterface::makeBlockSizeRange(m_exec_conf)},
+                                   this->m_exec_conf,
+                                   "cell_list"));
+    m_autotuners.push_back(m_tuner);
+
+    m_tuner_combine.reset(new Autotuner<1>({AutotunerInterface::makeBlockSizeRange(m_exec_conf)},
+                                           this->m_exec_conf,
+                                           "cell_list_combine"));
+    m_autotuners.push_back(m_tuner_combine);
```
Looking at this makes me wonder if we should use an unordered map. Then we wouldn't need to store autotuners in a member variable and `m_autotuners`.

joaander replied:
We could, yes. Would that set their string name as well? It would be the most logical connection between the Python dict of tuples interface. However if that were the case, how would they know their name to include in the notice messages?

b-butler replied:
Hmm, that is a good point. We could provide an interface for adding or removing tuners from the base class that set's it both places.

Otherwise, we could write an outside method/function that writes the logging messages (if possible I can't remember what messages are written).

joaander replied:
Though I do worry about the overhead of an `unordered_map` lookup every time we accesses one of these tuners. It seems unnecessarily complex. Is your goal to reduce the amount of code spent in initialization? I could collect all the ` m_autotuners.push_back` lines into one.

b-butler replied:
Yeah, if the notice messages are necessary then the name has to live in two places. I was trying to actually reduce the code complexity (from a human perspective not performance). Having to store every auto tuner in two places seems a bad pattern. We can't just use a `std::vector` though because that would obfuscate the code which is why I suggested an `unordered_map`. The slow down would only be while tuning correct? If so that shouldn't effect performance as a whole by much.

joaander replied:
> The slow down would only be while tuning correct? If so that shouldn't effect performance as a whole by much.

No, the tuner's `begin` and `end` methods are called for every kernel launch. The caller doesn't know whether the tuner is active or not.

b-butler commented on hoomd/System.cc:
```diff
@@ -188,31 +188,33 @@ void System::updateTPS()
-        m_integrator->setAutotunerParams(enabled, period);
+        m_integrator->startAutotuning();
+
+    //     //TODO: Integrator needs to call startAutotuning on all child classes, as do classes like
```
Comment highlighting the TODO

b-butler commented on sphinx-doc/features.rst:
```diff
@@ -77,6 +77,11 @@ run time with the `device <hoomd.device>` module. Unless otherwise stated in the
+Autotuned kernel launch parameters
+----------------------------------
+
+TODO
```
Comment marking todo

b-butler commented on hoomd/operation.py:
```diff
@@ -288,7 +288,88 @@ def __getstate__(self):
+
+    @property
+    def kernel_parameters(self):
+        """dict: Kernel parameters.
```
```suggestion
        """dict[str, tuple[float]]: Kernel parameters.
```

b-butler commented on hoomd/operations.py:
```diff
@@ -290,6 +291,41 @@ def computes(self):
+        for op in self:
+            result = result and op.is_tuning_complete
+
+        return result
```
```suggestion
        # TODO: Call C++ System object to do this, non-operations like
        # ParticleData, Communicator, etc... may be autotuned as well
        return all(op.is_tuning_complete for op in self)
```

b-butler reviewed (commented):
- :+1: to the new `Action` C++ class.
- I think making `m_autotuners` a `std::unordered_map<std::string, std::shared_ptr<AutotunerInterface>>` would make sense and we could remove the individual tuner member variables without losing readability from the name.
- I think the dict, tuple interface in Python is the right approach.

In general, I think the code make many good changes to the autotuner API both in C++ and Python. I have left specific comments in a few places. All the big level changes I have made above or commented in my review.

joaander committed:
Update hoomd/Autotuner.h

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Update hoomd/Autotuner.h

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Update hoomd/operation.py

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Update hoomd/operations.py

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Export autotuned methods in Autotuned.

joaander committed:
Merge branch 'trunk-minor' into refactor-autotuner

joaander committed:
Fix non-optional tuners complete after calling tune_kernel_parameters().

joaander committed:
Fix tune_kernel_parameters() call.

joaander committed:
Call shrink_to_fit.

joaander committed:
Rename median to center.

joaander committed:
Rename AutotunerInterface to AutotunerBase

joaander committed:
Add suggested comment.

joaander committed:
Merge branch 'trunk-minor' into refactor-autotuner

joaander committed:
Use insert to add multiple autotuners at once.

joaander committed:
Autotuners are not used in ParticleData or Communicator.

joaander committed:
Add pass-through support for tuners in md integrators.

joaander committed:
Make TOOD more recognizable.

joaander committed:
Add TODO's for pair potentials and nlist.

joaander committed:
Update Autotuners in md module.

joaander committed:
Run pre-commit.

joaander committed:
Update Autotuner API in mpcd package.

joaander committed:
Add virtual deconstructor.

joaander committed:
Pass on autotuning calls to cell list.

joaander committed:
Mark conditional tuners optional.

joaander committed:
Run pre-commit.

joaander committed:
Make forces and methods autotuned objects.

joaander committed:
Add overview documentation for autotuners.

joaander committed:
Reword features overview.

joaander committed:
Fix typo in features docs.

joaander committed:
Add unit tests.

Fix bugs found while testing.

joaander committed:
Fix unit tests.

joaander committed:
Merge branch 'trunk-minor' into refactor-autotuner

joaander committed:
Run pre-commit.

joaander committed:
Fix failing test.

joaander committed:
Skip reverseperturbation flow tests in MPI.

joaander committed:
Correctly set autotuner parameters for JIT patch kernels.

b-butler commented on hoomd/Autotuner.h:
```diff
@@ -24,227 +34,628 @@
+    getTppListPow2(const std::shared_ptr<const ExecutionConfiguration> exec_conf,
+                   int force_size = -1)
+        {
+        std::vector<unsigned int> v;
+        unsigned int warp_size = exec_conf->dev_prop.warpSize;
+        if (force_size > 0)
+            {
+            warp_size = static_cast<unsigned int>(force_size);
+            }
+
+        for (unsigned int s = 4; s <= warp_size; s *= 2)
+            {
+            v.push_back(s);
+            }
+        v.push_back(1);
+        v.push_back(2);
```
```suggestion
        std::vector<unsigned int> v {1, 2};
        unsigned int warp_size = exec_conf->dev_prop.warpSize;
        if (force_size > 0)
            {
            warp_size = static_cast<unsigned int>(force_size);
            }

        for (unsigned int s = 4; s <= warp_size; s *= 2)
            {
            v.push_back(s);
            }
```

joaander replied:
This ordering is intentional. Unit tests tend to trigger only the first few entries in the autotuner parameter list. By ensuring the more general multiple threads per particle path is the first executed the unit tests can find bugs in that code path. We made this change when the tpp=1 code path passed all unit tests but the tpp=2+ path failed. 

I'll add a comment to explain this reasoning.

b-butler commented on hoomd/hpmc/pytest/test_compute_free_volume.py:
```diff
@@ -76,6 +76,14 @@ def test_validation_systems(simulation_factory, two_particle_snapshot_factory,
+    # Tet the kernel parameter tuner.
+    def activate_tuner():
+        sim.run(1)
+        return free_volume_compute.free_volume
```
No need for the return statement
```suggestion
        # We need to make the kernel be called.
        free_volume_compute.free_volume
```

joaander replied:
I'll try this change. I didn't write it this way as I expected flake8 to complain.

b-butler commented on hoomd/md/ActiveForceComputeGPU.h:
```diff
@@ -31,22 +31,17 @@ class PYBIND11_EXPORT ActiveForceComputeGPU : public ActiveForceCompute
-        m_tuner_force->setEnabled(enable);
-        m_tuner_diffusion->setPeriod(period);
-        m_tuner_diffusion->setEnabled(enable);
+        // ActiveForceCompute::startAutotuning();
```
```suggestion
```

joaander replied:
Good catch. I should remove the whole method as this is just a reimplementation of `Autotuned::startAutotuning`.

b-butler commented on hoomd/md/PotentialPair.h:
```diff
@@ -843,6 +849,21 @@ CommFlags PotentialPair<evaluator>::getRequestedCommFlags(uint64_t timestep)
+    {
+    bool result = ForceCompute::isAutotuningComplete();
+    result = result && m_nlist->isAutotuningComplete();
+    return result;
```
```suggestion
    return result && m_nlist->isAutotuningComplete();
```

b-butler commented on hoomd/md/PotentialTersoff.h:
```diff
@@ -107,6 +107,23 @@ template<class evaluator> class PotentialTersoff : public ForceCompute
+        {
+        bool result = ForceCompute::isAutotuningComplete();
+        result = result && m_nlist->isAutotuningComplete();
+        return result;
```
```suggestion
        return result && m_nlist->isAutotuningComplete();
```

b-butler commented on hoomd/module.cc:
```diff
@@ -153,6 +154,18 @@ std::string mpi_bcast_str(pybind11::object string,
+#ifdef ENABLE_MPI
+    bool reduced_result = false;
+    MPI_Allreduce(&v, &reduced_result, 1, MPI_C_BOOL, MPI_LAND, exec_conf->getMPICommunicator());
+    bcast(reduced_result, 0, exec_conf->getMPICommunicator());
```
Shouldn't `MPI_Allreduce` handle this?
```suggestion
```

joaander replied:
Yes. The microsoft documentation (first hit on DuckDuckGo) is incorrect.

b-butler reviewed (commented):
Overall looks really good. I left a few comments.

joaander committed:
Apply suggestions from code review

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Remove unneeded bcast.

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Find compute-sanitizer on arch linux.

joaander committed:
Remove uneeded startAutotuning method.

joaander committed:
Run pre-commit.

joaander committed:
Quiet compiler warnings with gcc 12.

joaander commented:
#1374 documents some additional work to be completed in a future PR.

joaander committed:
Explain why elements 1 and 2 are at the end.

joaander committed:
Fix test_reverse_perturbation_flow.py

joaander committed:
Temporary debugging code.

Attempting to determine why hpmc/pytest/test_pair_union_user.py::test_kernel_parameters[GPU] fails in CI but not locally.

joaander committed:
Fix devices list for test.

joaander committed:
Run the proper test on CI.

joaander committed:
Remove testing code and run patch tests with more particles.

The additional particles ensure that the kernels run on all MPI ranks.

joaander committed:
Merge branch 'trunk-minor' into refactor-autotuner

joaander committed:
Fix broken merge.

joaander committed:
Attempt to fix compile errors with clang6.

joaander committed:
Find cuda-memcheck when building against CUDA 10.

joaander commented:
All tests passing. This is ready for the final review round.

Charlottez112 commented on hoomd/operation.py:
```diff
@@ -288,7 +288,95 @@ def __getstate__(self):
+        When you significantly change your system during the simulation (e.g.
+        compress to a higher density), then tune the parameters again after
+        making the change with `tune_kernel_parameters`.
+
```
This is a very helpful tip, but it's not very clear if this is something we always have to do after changing the system. I imagine it's necessary if the change significantly affects the performance?

joaander replied:
I added a statement that this is optional, but that it can increase performance.

Charlottez112 commented on hoomd/operation.py:
```diff
@@ -288,7 +288,95 @@ def __getstate__(self):
+        `AutotunedObject` reference the rank local tuner objects. Different
+        ranks may tune different optimal kernel parameters and may complete
+        tuning at different times.
+
```
It seems like this has some implications. Could you explain what's implied here?

joaander replied:
Users can inspect `is_tuning_complete` for individual objects, but should not use them to control the simulation - otherwise deadlock could result. I added a statement indicating that one should use `Operations.is_tuning_complete` for control flow.

Charlottez112 commented on ARCHITECTURE.md:
```diff
@@ -205,6 +207,19 @@ of bare pointers to data arrays, array sizes, etc... The ABI for these calls is
+It dynamically cycles through the possible parameters and records the performance of each using CUDA
+events. After scanning through all parameters, it selects the best performing one to continue
+executing. GPU code in HOOMD-blue should instantiate and use one `Autotuner` for each kernel.
+Classes that use the autotuner should inherit from `Autotuned` which tracks all the autotuners and
```
How does one know which class is autotuned?

joaander replied:
When a class has `Autotuner` member variables. I reworded the documentation to indicate this.

joaander committed:
Clarify documentation.

b-butler commented on hoomd/hpmc/pytest/test_pair_user.py:
```diff
@@ -89,14 +89,14 @@ def test_attaching(device, simulation_factory, two_particle_snapshot_factory):
 @pytest.mark.skipif(llvm_disabled, reason='LLVM not enabled')
-def test_kernel_parameters(simulation_factory, two_particle_snapshot_factory):
-    patch = hoomd.hpmc.pair.user.CPPPotential(r_cut=3,
+def test_kernel_parameters(simulation_factory, lattice_snapshot_factory):
```
Is there a reason for switching to use `lattice_snapshot_factory` in more places?

joaander replied:
With the two particle snapshot factory, one of the ranks had no particles and failed to run the kernel. Thus the autotuner on that rank didn't change and the assert checking for changed values failed.

b-butler reviewed (approved):
The changes look good to me.
